provider/core_security_list_resource.go

```diff
@@ -332,49 +322,94 @@ func (s *SecurityListResourceCrud) buildICMPOptions(conf map[string]interface{})
-	l := conf["tcp_options"].([]interface{})
-	if len(l) > 0 {
-		confOpts := l[0].(map[string]interface{})
+	options := conf["tcp_options"].([]interface{})
+	if len(options) > 0 {
+		sourcePortRange, destinationPortRange := s.buildSourceAndDestinationPortRanges(options)
 		opts = &baremetal.TCPOptions{
-			baremetal.PortRange{
-				Max: uint64(confOpts["max"].(int)),
-				Min: uint64(confOpts["min"].(int)),
-			},
+			destinationPortRange,
+			sourcePortRange,
 		}
 	}
 	return
 }
 
 func (s *SecurityListResourceCrud) buildUDPOptions(conf map[string]interface{}) (opts *baremetal.UDPOptions) {
-	l := conf["udp_options"].([]interface{})
-	if len(l) > 0 {
-		confOpts := l[0].(map[string]interface{})
+	options := conf["udp_options"].([]interface{})
+	if len(options) > 0 {
+		sourcePortRange, destinationPortRange := s.buildSourceAndDestinationPortRanges(options)
 		opts = &baremetal.UDPOptions{
-			baremetal.PortRange{
-				Max: uint64(confOpts["max"].(int)),
-				Min: uint64(confOpts["min"].(int)),
-			},
+			destinationPortRange,
+			sourcePortRange,
 		}
 	}
 	return
 }
 
-func buildConfICMPOptions(opts *baremetal.ICMPOptions) (list []interface{}) {
-	confOpts := map[string]interface{}{
-		"code": int(opts.Code),
-		"type": int(opts.Type),
+func buildPortRange(conf []interface{}) (portRange *baremetal.PortRange) {
+	if len(conf) > 0 && conf[0] != nil {
+		mapConf := conf[0].(map[string]interface{})
+
+		max := mapConf["max"].(int)
+		min := mapConf["min"].(int)
+
+		// Max and Min default to 0, and that is not a valid port number, so we can assume that if
+		// the value is 0 then the user has not set the port number.
+		// Also, note that if either max or min is set, then the service will return an error if both are not
+		// set. However, we want to create the PortRange if either is set and let the service return the error.
+		if max != 0 || min != 0 {
+			portRange = &baremetal.PortRange{
+				Max: uint64(max),
+				Min: uint64(min),
+			}
+		}
 	}
-	return []interface{}{confOpts}
+	return
 }
 
-func buildConfTransportOptions(portRange baremetal.PortRange) (list []interface{}) {
-	confOpts := map[string]interface{}{
-		"max": int(portRange.Max),
-		"min": int(portRange.Min),
+func (s *SecurityListResourceCrud) buildSourceAndDestinationPortRanges(conf []interface{}) (sourcePortRange, destinationPortRange *baremetal.PortRange) {
+	if len(conf) > 0 && conf[0] != nil {
+		mapConf := conf[0].(map[string]interface{})
+		sourcePortRange = buildPortRange(mapConf["source_port_range"].([]interface{}))
+		destinationPortRange = buildPortRange(conf)
 	}
-	return []interface{}{confOpts}
+
+	return
+}
+
+// Used to build rule lists for SetData.
+func buildConfRuleLists(res *baremetal.SecurityList) (confEgressRules, confIngressRules []map[string]interface{}) {
+	for _, egressRule := range res.EgressSecurityRules {
+		confEgressRule := map[string]interface{}{}
+		confEgressRule["destination"] = egressRule.Destination
+		confEgressRule = buildConfRule(
```

Comment:
The interface of the function `buildConfRule()` looks confusing because it takes a map (`confEgressRule` in this case) as input, then returns the map as it is, instead of a newly created map.  Probably `buildConfRule()` should create and return a new map as other `build*Options()` below.

Reply:
done